k2/csrc/default_context.cu

```diff
@@ -26,8 +26,8 @@ class CpuContext : public Context {
 
  public:
   static ContextPtr Make() {
-    auto p = new CpuContext();
-    return ContextPtr{p};
+    static ContextPtr ctx(new CpuContext());
```

Comment:
The destructor of `ContextPtr` is non-trivial.

According to https://google.github.io/styleguide/cppguide.html#Static_and_Global_Variables

> Objects with static storage duration are **forbidden** unless they are trivially destructible.

Reply:
The previous implementation (withou `static`) is not a real singleton pattern, but we may only need one instance for CPU context. That's why I change it to `static`.
 I just think we can not follow the google style here to make it a singleton, as seems no potential risk for our usages, what do you think? And I think if we must follow, we can just not use singleton pattern (I mean, the previous one which is a wrong implemetation), just make the default constructor public is fine.

Reply:
The original implementation is not a singleton design. It is used to ensure that you cannot create a raw pointer to
`Context` since we want to use its `shared_from_this` method.

Reply:
A workaround is to put the declaration and definition of the `xxxContext` in the `*.cu` files
and expose only `GetCpuContext()` and `GetCudaContext()`. Then we can make the constructors public
since they will be called only by us.

Reply:
That's a good idea, will update. BTW, according to google style, how do we do if we want a object constructed only once?

Reply:
Maybe `std::call_once`?

Reply:
mm...I doubt it as that may mean Google style only supports C++11 or later..

Reply:
https://google.github.io/styleguide/cppguide.html#C++_Version

> Currently, code should target C++17, i.e., should not use C++2x features. The C++ version targeted by this guide will advance (aggressively) over time.
> Consider portability to other environments before using features from C++14 and C++17 in your project.

Reply:
Even PyTorch requires a C++14 compiler.

Reply:
just wonder how they do this back to the time when most of compilers does not support C++11 perfectly. Anyway, updated (moved to `.cu`)

Reply:
[`pthread_once`][1] is an alternative for `std::call_once`.

[1]: https://pubs.opengroup.org/onlinepubs/007908799/xsh/pthread_once.html

Reply:
You can find its usage in the early versions of OpenFST:

https://android.googlesource.com/platform/external/srec/+/9164572ac412a36d61d05fe206a7c2811e7c8a65/tools/thirdparty/OpenFst/fst/lib/register.h#54

```cpp

static FstRegister<A> *GetRegister() {
  pthread_once(&register_init_, &FstRegister<A>::Init);
  return register_;
}
```